File: src/key.py
# Keys that will be played with the left hand.
left_keys = {
    # White Keys.
    'q': 'c',
    'w': 'd',
    'e': 'e',
    'r': 'f',
    't': 'g',
    'y': 'a',
    'u': 'b',

    # Black keys
    '2': 'c-sharp',
    '3': 'd-sharp',
    '5': 'f-sharp',
    '6': 'g-sharp',
    '7': 'a-sharp'
}

# Keys that will be played with the right hand.
right_keys = {
    # White Keys.
    'v': 'c',
    'b': 'd',
    'n': 'e',
    'm': 'f',
    ',': 'g',
    '.': 'a',
    '/': 'b',

    # Black keys
    'g': 'c-sharp',
    'h': 'd-sharp',
    'k': 'f-sharp',
    'l': 'g-sharp',
    ';': 'a-sharp'
}

octaves = {
    # Left hand octaves.
    'a': 1,
    's': 2,
    'd': 3,
    'f': 4,

    # Right hand octaves.
    'i': 5,
    'o': 6,
    'p': 7 
}


class KeyManager:
    def __init__(self):
        # Save previously entered items to the class.
        self.left = left_keys
        self.right = right_keys
        self.octaves = octaves
        self.valid_keys = list(self.left.keys()) + list(self.right.keys()) + list(self.octaves.keys())

        # Initial Octaves.
        self.left_octave = 4
        self.right_octave = 5

    def __getitem__(self, input):
        # Identifies entered input and returns the key along with octave that is associated with it.
        if(input == ' '):
            print("Shutting down piano")
            exit(0)
        if(self.valid_key(input)):
            if(input in list(self.octaves)):
                new_octave = self.octaves[input]
                if(new_octave >= 1 and new_octave <= 4):
                    self.left_octave = new_octave
                if(new_octave >= 5 and new_octave <= 7):
                    self.right_octave = new_octave
                return f"{new_octave} octave."
            if(input in list(self.left.keys())):
                return f"{self.left_octave}-{self.left[input]}"
            if(input in list(self.right.keys())):
                return f"{self.right_octave}-{self.right[input]}"
        else:
            return ""

    
    def valid_key(self, key):
        if(key not in self.valid_keys):
            return False
        else:
            return True

File: js/key.js
const left_key = new Map([
        // White Keys
        ['q', 'c'],
        ['w', 'd'],
        ['e', 'e'],
        ['r', 'f'],
        ['t', 'g'],
        ['y', 'a'],
        ['u', 'b'],
        // Black Keys
        ['2', 'c-sharp'],
        ['3', 'd-sharp'],
        ['5', 'f-sharp'],
        ['6', 'g-sharp'],
        ['7', 'a-sharp']
])
left = Array.from(left_key.keys())

const right_key = new Map([
        // White Keys
        ['v', 'c'],
        ['b', 'd'],
        ['n', 'e'],
        ['m', 'f'],
        [',', 'g'],
        ['.', 'a'],
        ['/', 'b'],
        // Black Keys
        ['g', 'c-sharp'],
        ['h', 'd-sharp'],
        ['k', 'f-sharp'],
        ['l', 'g-sharp'],
        [';', 'a-sharp']
])
right = Array.from(right_key.keys())

const octaves = new Map([
        // Left Hand Octaves.
        ['a', 1],
        ['s', 2],
        ['d', 3],
        ['f', 4],
        // Right Hand Octaves.
        ['i', 5],
        ['o', 6],
        ['p', 7]
])
const octaveKeys = Array.from(octaves.keys())
const validInput = octaveKeys.concat(left, right)

class KeyManager {
        
        constructor() {
                this.left = left_key
                this.right = right_key
                this.octaves = octaves
                this.validInput = validInput
                
                this.leftOctave = 4
                this.rightOctave = 5
        }

        getKey(input) {
                if(!validInput.includes(input)) {
                        return "Invalid Input"
                }
                if(Array.from(this.left.keys()).includes(input)) {
                        return this.leftOctave + "-" + this.left.get(input)
                }
                if(Array.from(this.right.keys()).includes(input)) {
                        return this.rightOctave + "-" + this.right.get(input)
                }
                if(Array.from(this.octaves.keys()).includes(input)) {
                        let newOctave = this.octaves.get(input)
                        if(newOctave >= 1 && newOctave <= 4) {
                                this.leftOctave = newOctave 
                        }
                        if(newOctave >= 5 && newOctave <= 7) {
                                this.rightOctave = newOctave
                        }
                        return newOctave + " octave"
                }
                return ""
        }

}
